### PR TITLE
Fix: Spring get bean of proxy OOM

### PR DESCRIPTION
**Pull request by zzhlhc**
As the issue #502  describes: Spring's `getbean(...)` method enhanced by `org.hotswap.agent.plugin.spring.getbean.ProxyReplacer` will repeatedly create the proxy class and its bean when obtaining the AOP-modified bean, so I cached the proxy bean to avoid repeated creation.

- **Let’s see the effect, Run the following code and compare the dumps:**

```
@Service
public class ServiceA {
    public String service() {
        return "" ;
    }
}

@Aspect
@Component
public class ServiceAAspect {
    @Pointcut("execution(public * fix.ServiceA.service(..)) "
    )
    public void pointCut_search() {
    }

    @Around("pointCut_search()")
    public Object around_search(ProceedingJoinPoint proceedingJoinPoint) throws Throwable {
        return proceedingJoinPoint.proceed();
    }
}

    public void getBean() {
        int i = 20;
        for (int i1 = 0; i1 < i; i1++) {
            applicationContext.getBean("serviceA");
        }
    }
```

- **Before fix:**

20 proxy classes were created repeatedly
5@49ZG@Q5RN8CUGC97ECR~F

- **After fix:**

only 1
V30 2ET%~1BFWM9414S7YBN

### CODE REVIEW

**Comment by zzhlhc** [comment]
@skybber @beiwei30 Hi, is anyone here to review this PR?

**Comment by skybber** [comment]
Please wait, we have to finish previous PR before look at your PR. The previous one is quite large. Anyway thanks for your work!

**Comment by skybber** [comment]
I've merged https://github.com/HotswapProjects/HotswapAgent/pull/496 and there are some conflicts with this PR now. Could you please fix them?

**Comment by zzhlhc** [comment]
> I've merged #496 and there are some conflicts with this PR now. Could you please fix them?

Sure, finished!

File: plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/reload/SpringBeanReload.java
Comment (cvictory) [inline plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/reload/SpringBeanReload.java:240]: I think maybe these clear action should be moved to destroy method of DefaultListableFactory. Because any bean is destroyed should clear the cache.

Comment (zzhlhc) [reply plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/reload/SpringBeanReload.java:240]: Okay, I'll see how to change it later.

**Comment by zzhlhc** [comment]
Thanks, it looks reasonable now, please check again.

**Comment by cvictory** [comment]
> Thanks, it looks reasonable now, please check again.

Hello， I try to find the better way to implements the Proxy. I think we can to optimize it :
* remove org.hotswap.agent.plugin.spring.transformers.ProxyReplacerTransformer#replaceBeanWithProxy method, instead we can enhance the method *org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory#createBean(java.lang.String, org.springframework.beans.factory.support.RootBeanDefinition, java.lang.Object[])* .

* As mentioned earlier, we should remove the cache when destroy method of BeanFactory is invoked.

what about this?

**Comment by zzhlhc** [comment]
> optimize

Glad to do some optimizations, let me look into it.

**Comment by chenghao111** [comment]
Hi, I have encountered the same issue you mentioned about Spring's "get bean of proxy" causing an OutOfMemoryError (OOM). In the previous version, I tested the fix code you provided, and it successfully resolved the performance problem. However, in the latest code after merging the refactoring changes, the issue is still not resolved. The OOM problem is occurring again, just like before.

**Comment by zzhlhc** [comment]
> Hi, I have encountered the same issue you mentioned about Spring's "get bean of proxy" causing an OutOfMemoryError (OOM). In the previous version, I tested the fix code you provided, and it successfully resolved the performance problem. However, in the latest code after merging the refactoring changes, the issue is still not resolved. The OOM problem is occurring again, just like before.

Hi, can you try my latest commit? I have tried it myself. If you still get OOM, please provide the reproduction steps and simple code. Thanks!

**Comment by zzhlhc** [comment]
@cvictory 
I made these two changes and the check of run-tests.sh passed, except `ConfigurationXmlTest#swapConfigClass()`, I think this is a problem with my IDEA, because even without my changes, this method still fails. So I think it is done.

Please check again.

File: plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/transformers/ProxyReplacerTransformer.java
Comment (cvictory) [inline plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/transformers/ProxyReplacerTransformer.java:51]: It's not necessary to loop all the method , maybe we just use ctClass.getDeclaredMethod(FACTORY_CREATE_METHOD_NAME, new CtClass[] {xxx}).

File: plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/transformers/ProxyReplacerTransformer.java
Comment (cvictory) [inline plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/transformers/ProxyReplacerTransformer.java:75]: we can just use getDeclaredMethod

File: plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/transformers/ProxyReplacerTransformer.java
Comment (zzhlhc) [reply plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/transformers/ProxyReplacerTransformer.java]: done

File: plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/transformers/ProxyReplacerTransformer.java
Comment (zzhlhc) [reply plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/transformers/ProxyReplacerTransformer.java]: done

**Comment by beiwei30** [comment]
I will try to reproduce why proxy classes can be leaked.

One comment after took a glance at this change:  It doesn't look right to me that ProxyReplacerTransformer starts to enhance createBean instead of getBean in the current approach. I think we should stick to enhance getBean so that org.hotswap.agent.plugin.spring.getbean.DetachableBeanHolder#getBean can cover all getBean scenarios.

**Comment by zzhlhc** [comment]
> I will try to reproduce why proxy classes can be leaked.
> 
> One comment after took a glance at this change: It doesn't look right to me that ProxyReplacerTransformer starts to enhance createBean instead of getBean in the current approach. I think we should stick to enhance getBean so that org.hotswap.agent.plugin.spring.getbean.DetachableBeanHolder#getBean can cover all getBean scenarios.

My Spring application usage scenarios are limited, but all tests have passed. If you find any missing scenarios, I can add test methods.

**Comment by beiwei30** [comment]
> > I will try to reproduce why proxy classes can be leaked.
> > One comment after took a glance at this change: It doesn't look right to me that ProxyReplacerTransformer starts to enhance createBean instead of getBean in the current approach. I think we should stick to enhance getBean so that org.hotswap.agent.plugin.spring.getbean.DetachableBeanHolder#getBean can cover all getBean scenarios.
> 
> My Spring application usage scenarios are limited, but all tests have passed. If you find any missing scenarios, I can add test methods.

No worry, I will come back once I understand the root cause better.

**Comment by beiwei30** [comment]
Today I made a fairly small change to solve this issue. It seems work too, @zzhlhc @cvictory Would you mind to take a look?

```diff
diff --git a/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java b/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java
index 80f2119d..1b666f66 100644
--- a/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java
+++ b/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java
@@ -25,7 +25,9 @@ import java.lang.reflect.Method;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
+import java.util.HashSet;
 import java.util.List;
+import java.util.Set;
 
 import org.hotswap.agent.logging.AgentLogger;
 
@@ -46,6 +48,7 @@ public class DetachableBeanHolder implements Serializable {
     private Object[] paramValues;
     private static List<WeakReference<DetachableBeanHolder>> beanProxies =
             Collections.synchronizedList(new ArrayList<WeakReference<DetachableBeanHolder>>());
+    private static Set<String> beanNames = Collections.synchronizedSet(new HashSet<>());
     private static AgentLogger LOGGER = AgentLogger.getLogger(DetachableBeanHolder.class);
 
     /**
@@ -58,14 +61,16 @@ public class DetachableBeanHolder implements Serializable {
      * @param paramValues
      */
     public DetachableBeanHolder(Object bean, Object beanFactry, Class<?>[] paramClasses, Object[] paramValues) {
-        if (bean == null) {
-            LOGGER.error("Bean is null. The param value: {}", Arrays.toString(paramValues));
-        }
         this.bean = bean;
         this.beanFactory = beanFactry;
         this.paramClasses = paramClasses;
         this.paramValues = paramValues;
         beanProxies.add(new WeakReference<DetachableBeanHolder>(this));
+        beanNames.add(bean.getClass().getName());
+    }
+
+    public static boolean hasRegistered(Object bean) {
+        return beanNames.contains((bean.getClass().getName()));
     }
 
     /**
diff --git a/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/ProxyReplacer.java b/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/ProxyReplacer.java
index c110416e..f2355ccc 100644
--- a/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/ProxyReplacer.java
+++ b/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/ProxyReplacer.java
@@ -62,7 +62,7 @@ public class ProxyReplacer {
      * @return Proxied bean
      */
     public static Object register(Object beanFactry, Object bean, Class<?>[] paramClasses, Object[] paramValues) {
-        if (bean == null) {
+        if (bean == null || DetachableBeanHolder.hasRegistered(bean)) {
             return bean;
         }
         if (SpringPlugin.basePackagePrefixes != null) {

```

**Comment by zzhlhc** [comment]
@beiwei30 
SpringPluginTest#hotswapPrototypeTestExistingInstance failed.
Can you explain in more detail why the previous changes were bad？

**Comment by zzhlhc** [comment]
> @beiwei30 SpringPluginTest#hotswapPrototypeTestExistingInstance failed. Can you explain in more detail why the previous changes were bad？

The reason for OOM is that there is no cache for instances similar to this class `BeanServiceImpl$HOTSWAPAGENT_$$EnhancerBySpringCGLIB$$bdb9e5d4_2@4503` in the Spring container.
The enhanced `createBean` method can cache the instance of `BeanServiceImpl$HOTSWAPAGENT_..` in the spring container, so I think enhanced createBean can fundamentally solve this problem.

**Comment by beiwei30** [comment]
> SpringPluginTest

I will investigate why it fails.

**Comment by beiwei30** [comment]
> @beiwei30 SpringPluginTest#hotswapPrototypeTestExistingInstance failed. Can you explain in more detail why the previous changes were bad？

As I said in my first comment on this thread, it doesn't looks good to me to change registration from 'getBean' to 'createBean'. The original impl is better because it registers on all 'getBean' methods, so that it can use the correct getBean method (found by paramClasses) and the correct arguments (paramValues) so that it can get the correct bean from getBean() method.

**Comment by zzhlhc** [comment]
> so that it can get the correct bean from getBean() method

The first parameter of the enhanced `createBean` method is `String beanName`. Each beanName has only one bean in the container, so as long as the beanName is correct, getBean(beanName) should be able to get the correct bean. 

Am I wrong?

**Comment by beiwei30** [comment]
> > so that it can get the correct bean from getBean() method
> 
> The first parameter of the enhanced `createBean` method is `String beanName`. Each beanName has only one bean in the container, so as long as the beanName is correct, getBean(beanName) should be able to get the correct bean.
> 
> Am I wrong?

good point, let me revisit your approach further.

**Comment by beiwei30** [comment]
@zzhlhc If unique bean is the fact, then I agree with the idea of enhancing at 'createBean' is more effective way to eliminate this particular issue. I refined your solution a little bit further. Pls. consider if it's reasonable.

```diff
diff --git a/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java b/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java
index 0e07c292..60f58ac4 100644
--- a/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java
+++ b/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java
@@ -40,11 +40,9 @@ public class DetachableBeanHolder implements Serializable {
 
     private static final long serialVersionUID = -7443802320153815102L;
 
+    private String beanName;
     private Object bean;
     private Object beanFactory;
-    private Class<?>[] paramClasses;
-    private Object[] paramValues;
-    //key: bean name
     private static final Map<String, List<WeakReference<DetachableBeanHolder>>> beanProxies = new ConcurrentHashMap<>();
     private static AgentLogger LOGGER = AgentLogger.getLogger(DetachableBeanHolder.class);
 
@@ -60,10 +58,8 @@ public class DetachableBeanHolder implements Serializable {
         }
         this.bean = bean;
         this.beanFactory = beanFactry;
-        this.paramClasses = paramClasses;
-        this.paramValues = paramValues;
 
-        String beanName = deduceBeanName(beanFactry, paramClasses, paramValues);
+        beanName = deduceBeanName(paramValues);
         if (beanName != null) {
             synchronized (beanProxies) {
                 beanProxies.computeIfAbsent(beanName, k -> new ArrayList<>());
@@ -72,21 +68,11 @@ public class DetachableBeanHolder implements Serializable {
         }
     }
 
-    protected static String deduceBeanName(Object beanFactory, Class<?>[] paramClasses, Object[] paramValues) {
-        String beanName = null;
-        if (String.class.getName().equals(paramClasses[0].getName())) {
-            beanName = paramValues[0].toString();
-        } else {
-            //getBeanNamesForType method will return only one here, because getBean with requiredType param has NoUniqueBeanDefinitionException check
-            String[] beanNamesForType = ((DefaultListableBeanFactory) beanFactory).getBeanNamesForType((Class<?>) paramValues[0]);
-            if (beanNamesForType.length == 1) {
-                beanName = beanNamesForType[0];
-            } else {
-                //this should never happen
-                LOGGER.warning("Method beanNamesForType return unexpected names:{}", String.join(",", beanNamesForType));
-            }
+    protected static String deduceBeanName(Object[] paramValues) {
+        if (String.class != paramValues[0].getClass()) {
+            throw new IllegalArgumentException("the first argument of doCreateBean method should be java.lang.String");
         }
-        return beanName;
+        return (String) paramValues[0];
     }
 
     /**
@@ -152,8 +138,6 @@ public class DetachableBeanHolder implements Serializable {
                 Class<?>[] parameterTypes = factoryMethod.getParameterTypes();
                 if (ProxyReplacer.FACTORY_METHOD_NAME.equals(factoryMethod.getName())
                         && parameterTypes.length == 1 && parameterTypes[0] == String.class) {
-
-                    String beanName = (String) paramValues[0];
                     Object freshBean = factoryMethod.invoke(beanFactory, beanName);
 
                     // Factory returns HA proxy, but current method is invoked from HA proxy!
@@ -167,10 +151,10 @@ public class DetachableBeanHolder implements Serializable {
                     bean = freshBean;
                     beanCopy = bean;
                     if (beanCopy == null) {
-                        LOGGER.debug("Bean of '{}' not loaded, {} ", bean.getClass().getName(), paramValues);
+                        LOGGER.debug("Bean '{}' not loaded, {} ", beanName);
                         break;
                     }
-                    LOGGER.info("Bean '{}' loaded", bean.getClass().getName());
+                    LOGGER.info("Bean '{}' loaded", beanName);
                     break;
                 }
             }
diff --git a/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/transformers/ProxyReplacerTransformer.java b/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/transformers/ProxyReplacerTransformer.java
index 2a513317..ec474677 100644
--- a/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/transformers/ProxyReplacerTransformer.java
+++ b/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/transformers/ProxyReplacerTransformer.java
@@ -29,7 +29,7 @@ import org.springframework.beans.factory.support.RootBeanDefinition;
  * @author Erki Ehtla
  */
 public class ProxyReplacerTransformer {
-    public static final String FACTORY_CREATE_METHOD_NAME = "createBean";
+    public static final String FACTORY_CREATE_METHOD_NAME = "doCreateBean";
 
     public static final String FACTORY_DESTROY_METHOD_NAME = "destroySingleton";
 

```

File: plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/transformers/ProxyReplacerTransformer.java
Comment (beiwei30) [inline plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/transformers/ProxyReplacerTransformer.java:32]: Enhancing 'doCreateBean' is better.

File: plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java
Comment (beiwei30) [inline plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java:26]: pls. avoid 'star import'.

File: plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java
Comment (beiwei30) [inline plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java:47]: instead of put a such simple comment here, I suggest to explain why a map is necessary, especially why the value is a list.

File: plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java
Comment (beiwei30) [inline plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java:75]: If the first argument is always bean name, then the logic of 'deduceBeanName' can be simplified.

**Review by beiwei30 — CHANGES_REQUESTED** [review]
@zzhlhc I've dropped my review comments.

**Comment by zzhlhc** [comment]
> @zzhlhc I've dropped my review comments.

Thanks for your reviews, I will look into them this weekend.

File: plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/transformers/ProxyReplacerTransformer.java
Comment (zzhlhc) [reply plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/transformers/ProxyReplacerTransformer.java:32]: Please see `SpringPluginTest#hotswapBeanCreatedByIABPP` i just added ,this method complements the test scenario and proves that `doCreateBean` is not the right choice. 

You can change `ProxyReplacerTransformer#FACTORY_CREATE_METHOD_NAME` to `doCreateBean` and execute this test method. The key point of why failed is that the few lines of code above `doCreateBean`
```
try {
	// Give BeanPostProcessors a chance to return a proxy instead of the target bean instance.
	Object bean = resolveBeforeInstantiation(beanName, mbdToUse);
	if (bean != null) {
		return bean;
	}
}
```
 will directly return the `SpringCglibxxx` bean in some cases, instead of the `HOTSWAPAGENT$SpringCglibxxx` bean.

**Comment by zzhlhc** [comment]
@beiwei30 
All finished except Enhancing 'doCreateBean', please check again.

File: plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java
Comment (beiwei30) [inline plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java:65]: Object beanName should be 'String beanName', and beanName can never be null.

```diff
-    public DetachableBeanHolder(Object bean, Object beanFactory, Object beanName) {
+    public DetachableBeanHolder(Object bean, Object beanFactory, String beanName) {
         String beanNameStr = beanName.toString();
         this.bean = bean;
         this.beanFactory = beanFactory;
         this.beanName = beanNameStr;
 
-        if (beanName != null) {
-            synchronized (beanProxies) {
-                beanProxies.computeIfAbsent(beanNameStr, k -> new ArrayList<>());
-                beanProxies.get(beanName).add(new WeakReference<>(this));
-            }
+        synchronized (beanProxies) {
+            beanProxies.computeIfAbsent(beanNameStr, k -> new ArrayList<>());
+            beanProxies.get(beanName).add(new WeakReference<>(this));
         }
     }

```

File: plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/transformers/ProxyReplacerTransformer.java
Comment (beiwei30) [reply plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/transformers/ProxyReplacerTransformer.java:32]: I see. Thanks for pointing it out.

**Review by beiwei30 — CHANGES_REQUESTED** [review]
@zzhlhc I have no further comment except one against DetachableBeanHolder's constructor.

File: plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java
Comment (zzhlhc) [reply plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java]: done

Comment (zzhlhc) [reply plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java]: done

**Comment by beiwei30** [comment]
@cvictory would you mind to take a look at @zzhlhc 's change too?

File: plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java
Comment (cvictory) [inline plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java:138]: One last question, If we use beanFactory.getBean(xxx.class), and the name of this bean is changed, it will cannot get the right bean.   Maybe we should store the getBean method and args.

I have no other question.

Comment (zzhlhc) [reply plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java:138]: You are right, but now we are enhancing `createBean` instead of `getBean`, so it is not so easy to get the parameters of `getBean`.

Regarding the way to get parameters, what I can think of is to enhance `getBean`, store the method parameters in `ThreadLocal`, and take them out when creating `DetachableBeanHolder`. What do you think? @cvictory @beiwei30

**Comment by beiwei30** [comment]
@zzhlhc @cvictory 

Today I revisit this issue to figure out if there's a better way to solve it. I noticed that there's a piece of code in EnhancrProxyCreater is commented out, and realized it should serve the purpose of avoiding repeatedly creating proxy classes. 

Once I revert it back, it does solve the issue we are working here:

```diff
diff --git a/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/EnhancerProxyCreater.java b/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/EnhancerProxyCreater.java
index 55b28e0d..fb20ec28 100644
--- a/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/EnhancerProxyCreater.java
+++ b/plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/EnhancerProxyCreater.java
@@ -18,7 +18,13 @@
  */
 package org.hotswap.agent.plugin.spring.getbean;
 
-import org.hotswap.agent.javassist.*;
+import org.hotswap.agent.javassist.CannotCompileException;
+import org.hotswap.agent.javassist.ClassPool;
+import org.hotswap.agent.javassist.CtClass;
+import org.hotswap.agent.javassist.CtMethod;
+import org.hotswap.agent.javassist.CtNewMethod;
+import org.hotswap.agent.javassist.LoaderClassPath;
+import org.hotswap.agent.javassist.NotFoundException;
 import org.hotswap.agent.logging.AgentLogger;
 import org.springframework.core.SpringVersion;
 
@@ -94,29 +100,28 @@ public class EnhancerProxyCreater {
     }
 
     private Object create(Object beanFactry, Object bean, Class<?>[] paramClasses, Object[] paramValues) {
-        return doCreate(beanFactry, bean, paramClasses, paramValues);
-//        Object proxyBean = null;
-//        if (beanProxies.containsKey(bean)) {
-//            proxyBean = beanProxies.get(bean);
-//        } else {
-//            synchronized (beanProxies) {
-//                if (beanProxies.containsKey(bean)) {
-//                    proxyBean = bean;
-//                } else {
-//                    proxyBean = doCreate(beanFactry, bean, paramClasses, paramValues);
-//                }
-//                beanProxies.put(bean, proxyBean);
-//            }
-//        }
-//
-//        // in case of HA proxy set the target. It might be cleared by clearProxies
-//        //   but the underlying bean did not change. We need this to resolve target bean
-//        //   in org.hotswap.agent.plugin.spring.getbean.DetachableBeanHolder.getBean()
-//        if (proxyBean instanceof SpringHotswapAgentProxy) {
-//            ((SpringHotswapAgentProxy) proxyBean).$$ha$setTarget(bean);
-//        }
-//
-//        return proxyBean;
+        Object proxyBean = null;
+        if (beanProxies.containsKey(bean)) {
+            proxyBean = beanProxies.get(bean);
+        } else {
+            synchronized (beanProxies) {
+                if (beanProxies.containsKey(bean)) {
+                    proxyBean = bean;
+                } else {
+                    proxyBean = doCreate(beanFactry, bean, paramClasses, paramValues);
+                }
+                beanProxies.put(bean, proxyBean);
+            }
+        }
+
+        // in case of HA proxy set the target. It might be cleared by clearProxies
+        //   but the underlying bean did not change. We need this to resolve target bean
+        //   in org.hotswap.agent.plugin.spring.getbean.DetachableBeanHolder.getBean()
+        if (proxyBean instanceof SpringHotswapAgentProxy) {
+            ((SpringHotswapAgentProxy) proxyBean).$$ha$setTarget(bean);
+        }
+
+        return proxyBean;
     }
 
     private Object doCreate(Object beanFactry, Object bean, Class<?>[] paramClasses, Object[] paramValues) {
```

This code is commented out in change a55c476d, and according to the commit log, it tries to solve issue #232 for spring 4.2.6 and 4.3.0. I tried to run all unit tests against 4.2.6 and 4.3.0 and all tests can pass.

I also verified this solution does solve the memory issue with the help from alibaba/arthas:

```shell
Connected to localhost.
Escape character is '^]'.
  ,---.  ,------. ,--------.,--.  ,--.  ,---.   ,---.
 /  O  \ |  .--. ''--.  .--'|  '--'  | /  O  \ '   .-'
|  .-.  ||  '--'.'   |  |   |  .--.  ||  .-.  |`.  `-.
|  | |  ||  |\  \    |  |   |  |  |  ||  | |  |.-'    |
`--' `--'`--' '--'   `--'   `--'  `--'`--' `--'`-----'

wiki       https://arthas.aliyun.com/doc
tutorials  https://arthas.aliyun.com/doc/arthas-tutorials.html
version    3.7.1
main_class com.baeldung.proxy.AspectApp
pid        79437
time       2023-11-17 20:50:20

[arthas@79437]$ sc *HOTSWAP*
com.baeldung.proxy.ServiceA$HOTSWAPAGENT_$$EnhancerBySpringCGLIB$$7471c848
com.baeldung.proxy.ServiceA$HOTSWAPAGENT_$$EnhancerBySpringCGLIB$$7471c848$HOTSWAPAGENT_$$FastClassBySpringCGLIB$$1a0ba636
com.baeldung.proxy.ServiceA$HOTSWAPAGENT_$$FastClassBySpringCGLIB$$efb767dd
Affect(row-cnt:3) cost in 40 ms.
```

I suggest we should continue to work on this approach. At the same time, reopen #232 to investigate further.  What do you think? 

@skybber pls. kindly put your comments too :)

**Comment by skybber** [comment]
What can I say as Jakarta user... , great work and I admire your enthusiasm.  Is it PR finished, can I merge it?

**Comment by zzhlhc** [comment]
This issue has gone on for too long. I hope to solve it quickly, any way is ok for me as long as it can solve this problem. 
Just let me know what needs to be done after you reviewers make the final decision.

**Comment by zzhlhc** [comment]
> ```diff
> beanProxies.put(bean, proxyBean)
> ```

BTW, it uses inside bean as key , good point.

**Comment by cvictory** [comment]
I think we check @beiwei30 ‘s comments next week. if it solved the issue, it is the simplest way.

**Comment by beiwei30** [comment]
@zzhlhc Pls. try the solution I proposed at https://github.com/HotswapProjects/HotswapAgent/pull/503#issuecomment-1816383871. If it works on your side too, I encourage you to continue your work with this approach.

**Comment by zzhlhc** [comment]
> try the solution I proposed at [#503 (comment)](https://github.com/HotswapProjects/HotswapAgent/pull/503#issuecomment-1816383871). If it works on your side too

Finished, and everything works fine.

**Comment by skybber** [comment]
Great, thanks!

**Comment by skybber** [comment]
@beiwei30: Would you please have time to look at #504 as it modifies the Spring plugin?

**Comment by beiwei30** [comment]
> @beiwei30: Would you please have time to look at #504 as it modifies the Spring plugin?

sure, will do.

**Pull request merged**
